Molecular Hamiltonian

lch opened the pull request:
## New features

This PR addresses issue #321 by introducing the `MolecularHamiltonian` type along with its related types and methods. These additions enable `Rimu.jl` to be used for post–Hartree-Fock calculations in quantum chemistry research accompanied with [`ElemCo.jl`](https://elem.co.il/).

A detailed description to the PR could be find at #348 .

lch committed:
MolecularHamiltonian (#321)

Add initial skeleton implementation for MolecularHamiltonian type

lch committed:
MolecularHamiltonian (#321)

Add initial implementation of `diagonal_element` for Molecular Hamiltonian

lch committed:
Molecular Hamiltonian (#321)

1. Avoid hard-coded data types; use a type that can be specified or inferred from FCIDUMP.
2. Fix the constructor to correctly calculate the number of electrons.

lch committed:
Molecular Hamiltonian (#321)

Move the `diagonal_element` implementation into `operator_column`

lch committed:
Molecular Hamiltonian (#321)

Store `occ_modes` in `MolecularHamiltonianOperatorColumn` to avoid recalculating it repeatedly.

lch committed:
Molecular Hamiltonian (#321)

Add an implementation of the num_offdiagonals method to the MolecularHamiltonianOperatorColumn

lch committed:
Molecular Hamiltonian (#321)

Export MolecularHamiltonian and MolecularHamiltonianOperatorColumn

lch committed:
Molecular Hamiltonian (#321)

Initial attempt at implementing offdiagonal iterator.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,201 @@
+    unoccupied::Vector{Vector{T}}
+end
+
+function modes_extract(addr::FermiFS2C)
+    occupied_modes = Vector{Vector{FermiFSIndex}}(undef, 2)
+    unoccupied_modes = Vector{Vector{FermiFSIndex}}(undef, 2)
+    for (i, state) in enumerate(addr.components)
+        occupied_modes[i] = OccupiedModeMap(state)
+        state_unoccupied_modes = FermiFSIndex[]
+        for j = 1:num_modes(state)
+            ind = find_mode(state, j)
+            if ind.occnum == 0
+                push!(state_unoccupied_modes, ind)
+            end
+        end
+        unoccupied_modes[i] = state_unoccupied_modes
+    end
+    Modes(occupied_modes, unoccupied_modes)
+end
```
As discussed, this look similar to what `OccupiedModeMap` is already doing. It would be better to extend the code in `src/BitstringAddresses/fockaddress.jl` to cater for `CompositeFS` addresses, and possibly for empty modes as well if that is required. It should be done using similar techniques as used for `OccupiedModeMap` including `SVector`s or `Tuple`s rather than `Vector`s to avoid heap allocations.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,201 @@
+            end
+        end
+    end
+    return MolecularHamiltonianOffDiagonals(mhod.modes, [cur_trans]), state
```
According to the interface `offdiagonals` should return an iterator over `Pair`s with a new address and the value of the matrix element. Thus, the `iterate` function needs to return a tuple of the form `(address => value, state)`, or `nothing` when it is exhausted.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,201 @@
+    n_one_electron_excitation + n_two_electron_excitation
+end
+
+const ModeIndex = @NamedTuple{ch::Int, orb::Int}
+
+struct ModeTransition
+    old::ModeIndex
+    new::ModeIndex
+end
```
Just reading this code I'm not sure I understand what it is meant to represent. What do `ch` and `orb` stand for? I'd like to encourage you to write descriptive docstrings right away.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,201 @@
+    new::ModeIndex
+end
+
+mutable struct MolecularHamiltonianOffDiagonalsIterState
+    occ_ind::Int
+    unocc_ind::Int
+    spin_ch::Int
+end
```
I think @mtsch provided an example for an iterator without a mutable state. The problem with `mutable struct` is that they cause heap allocations as well.

lch committed:
Molecular Hamiltonian (#321)

1. Simplified the modes_extract function using the new UnoccupiedModeMap interface.
2. Corrected the num_offdiagonals function.

joachimbrand commented on src/BitStringAddresses/fermifs.jl:
```diff
@@ -131,6 +131,22 @@ Base.:(==)(a::FermiFS, b::FermiFS) = a.bs == b.bs
+num_unoccupied_modes(::FermiFS{N,M}) where {N,M} = M - N
+unoccupied_modes(a::FermiFS{N,<:Any,S}) where {N,S} = FermiUnoccupiedModes{N,S}(~a.bs)
+
+function UnoccupiedModeMap(addr::FermiFS{N,M}) where {N,M}
```
In Julia we use CamelCase typically only for types. Functions can be CamelCase when they return a type of the same name (i.e. are type constructors). Here you have a CamelCase'd function that returns a type with a different name, and we should avoid this.

As a suggested resolution, just use snake_case for the name of the function:
```suggestion
function unoccupied_mode_map(addr::FermiFS{N,M}) where {N,M}
```
It is still a bit weird that your function returns a type `OccupiedModeMap`, which however stores a map of the unoccupied modes. Maybe we should rename  `OccupiedModeMap` into `ModeMap`, as it can hold any type of modes (occupied or not)?

lch committed:
Molecular Hamiltonian (#321)

Use concrete type in Modes struct.

lch committed:
Molecular Hamiltonian (#321)

Finish the MolecularHamiltonianOperatorColumn interface.

lch committed:
Molecular Hamiltonian (#321)

Resolve test failures in the test_hamiltonian_interface testset.

lch committed:
Molecular Hamiltonian (#321)

1. Use the default implementation of the random_offdiagonals function.
2. Align field naming in MolecularHamiltonianOperatorColumn struct.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,262 @@
+function ref_num_offdiagonals(column::MolecularHamiltonianOperatorColumn)
+    n_orb = num_modes(column.address.components[1])
+    n_alpha_elec = num_occupied_modes(column.address.components[1])
+    n_beta_elec = num_occupied_modes(column.address.components[2])
```
Use `num_particles` here

joachimbrand reviewed (commented):
Just another small comment. Otherwise this looks very promising.

joachimbrand commented:
Had a quick look at the failing tests: This might be the same issue @mtsch saw recently and fixed on the `develop` branch. If you merge `develop` it should be resolved.

lch committed:
Molecular Hamiltonian (#321)

Using num_particles in ref_num_offdiagonals

lch committed:
Molecular Hamiltonian (#321)

1. Eliminate memory allocations in unoccupied_mode_map
2. Rename UnoocupiedModeMap struct

lch committed:
Molecular Hamiltonian (#321)

Update calling functions according to to PR#330

lch committed:
Molecular Hamiltonian (#321)

1. Updated the usage of occupied_mode_map.
2. Moved the test to a separate file for better organization.

lch committed:
Molecular Hamiltonian (#321)

Fix offdiagonal term

lch committed:
Molecular Hamiltonian (#321)

Add tests for MolecularHamiltonian

lch committed:
Molecular Hamiltonian (#321)

1. Add H2 molecule FCIDUMP file for tests
2. Add test for ground state energy solution

mtsch commented on test/molecular_hamiltonian_tests.jl:
```diff
@@ -0,0 +1,44 @@
+            0 0.1809125853 -0.3437220243 0;
+            0.1809125853 0 0 0.4746278719
+        ]
+        fcidump = string(@__DIR__, "/examples/h2.FCIDUMP")
```
```suggestion
        fcidump = joinpath(@__DIR__, "/examples/h2.FCIDUMP")
```
What you wrote is not wrong, but using `joinpath` is preferred when dealing with paths. It makes your intention clearer and I think it handles path separators on Windows for you.

lch committed:
Molecular Hamiltonian (#321)

1. Use joinpath for portability
2. Merge from develop

lch committed:
Molecular Hamiltonian (#321)

1. Add starting_address argument for constructor of
   MolecularHamiltonian.
2. Add tests for modified constructor.
3. Fix path string of example FCIDUMP file.

lch committed:
Molecular Hamiltonian (#321)

Fix calculation results

lch committed:
Molecular Hamiltonian (#321)

Implement iterator-based offdiagonal term generation

lch committed:
Molecular Hamiltonian (#321)

Fix type issue

lch committed:
Molecular Hamiltonian (#321)

Add tests for offdiagonal term iterator

lch committed:
Molecular Hamiltonian (#321)

Add tests for offdiagonal term iterator

lch committed:
Molecular Hamiltonian (#321)

1. Remove mutable struct
2. Remove ods storage in MolecularHamiltonianOperatorColumn

lch committed:
Molecular Hamiltonian (#321)

Eliminate memory allocations in offdiagonal term generation.

lch committed:
Molecular Hamiltonian (#321)

Eliminate memory allocations in offdiagonal term generation.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -186,21 +164,21 @@ function ModeIndex{2}()
-    address::FermiFS2C
-    op::MolecularHamiltonian
-    modes::Modes
+struct MolecularHamiltonianOffDiagonalsIterator{NA,NB,T,A<:FermiFS2C,D,OA,OB,UA,UB,TI<:FermiFSIndex}
+    address::A
+    op::MolecularHamiltonian{T,A,D}
+    modes::Modes{OA,OB,UA,UB,TI}
```
There are now so may type parameters that the code becomes hard to read. Maybe this can be simplified, while still maintaining type stability?
```suggestion
struct MolecularHamiltonianOffDiagonalsIterator{NA,NB,T,A<:FermiFS2C,H<:MolecularHamiltonian{T,A},M<:Modes}
    address::A
    op::H
    modes::M
```
If you still want to check that `Modes` is sensible, this can be done in the type constructor.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,388 @@
+    end
+end
+
+struct ModeIndex{N}
+    index::NTuple{N,Int64}
+end
```
I don't understand what `ModeIndex` is doing or what it is good for. Please add docstrings also for internal types or functions to help the reader (unless your code is very simple and self-documenting).

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,388 @@
+
+function num_excitation(::MolecularHamiltonianOffDiagonalsIterator{NA,NB}) where {NA,NB}
+    return (NA, NB)
+end
```
What is `NA` and `NB`? I thought that this would refer to the number of alpha and beta electrons. Is it the number of excitations for alphas and betas instead? Please document this more clearly!

lch committed:
Molecular Hamiltonian (#321)

Fix bugs and add tests.

lch committed:
Molecular Hamiltonian (#321)

Improve type stability.

lch committed:
Molecular Hamiltonian (#321)

1. Add docstring for MolecularHamiltonian and functions
2. Fix typo in flip_spin_chan function

lch committed:
Molecular Hamiltonian (#321)

1. Fix and improve docstring
2. Prettify the code

joachimbrand commented on test/Hamiltonians.jl:
```diff
@@ -12,6 +12,8 @@ using Rimu.InterfaceTests: test_observable_interface, test_operator_interface,
     AdjointUnknown
 
+using ElemCo
+
```
```suggestion
```
This should not be needed any more, as testing the `MolecularHamiltonian` is done in a separate file.

joachimbrand commented on src/Hamiltonians/Hamiltonians.jl:
```diff
@@ -86,6 +88,8 @@ export hubbard_dispersion, continuum_dispersion
 export FroehlichPolaron
 export ParticleNumberOperator
 
+export MolecularHamiltonian, MolecularHamiltonianOperatorColumn
```
```suggestion
export MolecularHamiltonian
```
I don't think the column struct needs to be exported. The user shouldn't have to deal with it.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    unoccupied::Tuple{ModeMap{UA,T},ModeMap{UB,T}}
+end
+
+function modes_extract(addr::FermiFS2C)
```
I suggest a rename in analogy to `occupied_mode_map` and `unoccupied_mode_map`.
```suggestion
function mode_maps(addr::FermiFS2C)
```
Alternatively, and more verbose, `all_mode_maps` or `un_and_occupied_mode_maps`?

This function could later be used as a general constructor for different types of addresses. For this reason I suggested the renaming of the `Modes` struct to something more specific. This function should also live in `BitstringAddresses/`.

Please also add a docstring (but there is no need to export the symbol)!

lch replied:
I have renamed the `ModeMap` to `FermiFS2CModes`, renamed `modes_extract` to `full_mode_maps`, and move them to `BitStringAddresses` module.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    address::A
+    op::MolecularHamiltonian{T,A,D}
+    modes::Modes{OA,OB,UA,UB,TI}
+end
```
A suggestion to rename (simplify the name) and declutter the type parameters:
```suggestion
struct MolecularHamiltonianOffDiagonals{
    T,A<:FermiFS2C,H<:MolecularHamiltonian{T,A},M<:Modes
}
    address::A
    op::H
    modes::M
end
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    addr::A, op::MolecularHamiltonian{T,A,D}, m::Modes{OA,OB,UA,UB,TI}
+) where {T,A,D,OA,OB,UA,UB,TI}
+    return MolecularHamiltonianOffDiagonalsIterator{T,A,D,OA,OB,UA,UB,TI}(addr, op, m)
+end
```
```suggestion
```
I don't think that you need this method, because the Julia compiler already knows how to construct the type if you pass the three field arguments.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+function Base.eltype(
+    ::MolecularHamiltonianOffDiagonalsIterator{T,A}
+) where {T,A}
+    return Tuple{A,T}
```
```suggestion
    return Pair{A,T}
```
Please make it iterator over `Pair`s!

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    n_excited::Tuple{Int,Int}
+    from::Tuple{Int,Int}
+    to::Tuple{Int,Int}
+end
```
I suggest renaming for clarity, as "self-documenting code" (check if correct!):
```suggestion
struct MolecularHamiltonianOffDiagonalsIteratorState
    excitations_per_channel::Tuple{Int,Int} # either 0, 1, 2
    from_occupieds::Tuple{Int,Int} # indices into ModeMap
    to_unoccupieds::Tuple{Int,Int} # indices into ModeMap
end
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+struct Modes{OA,OB,UA,UB,T<:FermiFSIndex}
+    occupied::Tuple{ModeMap{OA,T},ModeMap{OB,T}}
+    unoccupied::Tuple{ModeMap{UA,T},ModeMap{UB,T}}
+end
```
I suggest a rename here, as well as switching the order of the type parameters. The typeindex `TI` is meaningful and thus should go first (in case one wants to check what type of indices it works with). The others are dummys and less meaningful.

```suggestion
struct FermiFS2CModeMaps{TI<:FermiFSIndex,OA,OB,UA,UB}
    occupied::Tuple{ModeMap{OA,TI},ModeMap{OB,TI}}
    unoccupied::Tuple{ModeMap{UA,TI},ModeMap{UB,TI}}
end
```
This type should live in `BitStingAddresses/multicomponent.jl`.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+end
+
+function Base.iterate(iter::MolecularHamiltonianOffDiagonalsIterator)
+    initial = MolecularHamiltonianOffDiagonalsIteratorState(0, 1)
```
```suggestion
    initial = MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (1, 0), (1, 0))
    # start with single electron excitation in beta channel
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    state::MolecularHamiltonianOffDiagonalsIteratorState
+)
+    na, nb = state.n_excited
+    if (na, nb) == (0, 1)
```
```suggestion
    if (na, nb) == (0, 1) # single electron in beta channel
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+        r = one_electron_excitation(2, iter.address, iter.op, iter.modes, state)
+        nstate = one_electron_excitation_next(iter, state)
+        return r, nstate
+    elseif (na, nb) == (1, 0)
```
```suggestion
    elseif (na, nb) == (1, 0) # single electron excitation in alpha channel
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+        r = one_electron_excitation(1, iter.address, iter.op, iter.modes, state)
+        nstate = one_electron_excitation_next(iter, state)
+        return r, nstate
+    elseif (na, nb) == (0, 2)
```
```suggestion
    elseif (na, nb) == (0, 2) # 2 electron excitation in beta channel
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+        r = two_electron_excitation(2, iter.address, iter.op, iter.modes, state)
+        nstate = two_electron_excitation_next(iter, state)
+        return r, nstate
+    elseif (na, nb) == (2, 0)
```
```suggestion
    elseif (na, nb) == (2, 0) # 2 electron excitation in alpha channel
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    is_void_state(s::MolecularHamiltonianOffDiagonalsIteratorState)
+
+"""
+function is_void_state(s::MolecularHamiltonianOffDiagonalsIteratorState)
```
In which case are void states produced? Do we really need this function?

The docstring is incomplete.

lch replied:
The "void" state serves as the `nothing` value returned by the sub-iterators. This ensures that the sub-iteration consistently returns a single type, thereby avoiding type instability.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    )
+end
+
+function MolecularHamiltonianOffDiagonalsIteratorState(na::Int, nb::Int)
```
It would be better to avoid such functions that have to check many `if` conditions on the passed arguments as this will cause "branching" code, which can be slow (e.g. if the branch has to jump to a code page that is not already loaded in the cache memory).

I don't think this is needed, though, as you can use the full constructor every time. It will also be easier to read.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    if (na, nb) == (0, 1)
+        if is_void_state(state) || is_invalid_state(iter, state)
+            return iterate(
+                iter, MolecularHamiltonianOffDiagonalsIteratorState(1, 0)
```
```suggestion
                iter, MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (1, 0), (1, 0))
```
See comment above!

Please apply to all the other cases as well!

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    ods = offdiagonals(column)
+    r = rand(collect(ods))
+    return r[1], 1 / length(ods), r[2]
+end
```
This function allocates because of the call to `collect`. While this can be easily avoided (because you implemented `length(offdiagonals(column))`), it will still be slow and inefficient.
I assume this is work in progress, but the final implementation should look quite similar to the one of the `offdiagonals` iterator.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+)
+    na, nb = state.n_excited
+    if (na, nb) == (0, 1)
+        if is_void_state(state) || is_invalid_state(iter, state)
```
These calls to the two functions look like they will generate a lot of branching code with many `if` conditions that need to be checked. It should be sufficient here to check whether the single electron excitations in the beta channel are exhausted. 
```suggestion
        if state.to[1] >  length(iter.modes.unoccupied[2])
```
or similar

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+This function is used to check if `s` is a valid state. It checks if the `from` and
+`to` are in the range of corresponding `ModeMap` array indices.
+"""
+function is_invalid_state(
```
As also said elsewhere, it is better to avoid writing code that goes through too many if conditions. See suggestions for how to do this below.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    elseif chan == 2
+        return 1
+    end
+end
```
This can be implemented without `if` condition to avoid branching code:
```julia
@inline flip_spin_chan(chan::Int) = 3 - chan
```
I also have a comment about naming. Please avoid abbreviations like "chan" in function names!
More generally, to be consistent with the rest of the code, I'd prefer to talk about "spin component" rather than "spin channel":
```suggestion
@inline flip_spin_component(component::Int) = 3 - component
```

joachimbrand replied:
Instead of making this an inlined function, you could also replace the function call by `3 - chan` wherever you used it. It is still quite good for readability (and speed).

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    
+    MolecularMolecularHamiltonian(fd::ElemCo.FciDumps.FDump[, starting_address::FermiFS2C])
+
+Implements a Molecular Hamiltonian with [ElemCo.jl](https://elem.co.il). 
```
```suggestion
    MolecularHamiltonian(fcidump::String[, starting_address::FermiFS2C])
    MolecularHamiltonian(fd::ElemCo.FciDumps.FDump[, starting_address::FermiFS2C])

Implements an electronic ab-initio Hamiltonian based on electron overlap integrals in
FCIDUMP format. It can be used to describe the electronic structure of a molecule
with fixed nuclear positions (i.e. under the Born-Oppenheimer approximation).
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+    \\hat{H}   & = \\hat{H}_0 + \\hat{H}_1 + \\hat{H}_2, \\\\
+    \\hat{H}_1 & = \\sum_{i,j} h_{ij} a^{\\dagger}_j a_i, \\\\
+    \\hat{H}_2 & = \\sum_{kl,ij} V_{kl,ij} a^{\\dagger}_{k} a^\\dagger_{l} a_j a_i.
+\\end{aligned}
```
The formula is missing the spin indices. Please correct!

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,613 @@
+!!! warning
+    The FCIDUMP file is assumed to be a Restricted Hartree-Fock(RHF) Hamiltonian.
+"""
+struct MolecularHamiltonian{T,A<:FermiFS2C,D<:FDump} <: AbstractHamiltonian{T}
+    starting_address::A
+    fcidump::D
+end
```
It would be nice to have a method for `show(io, h::MolecularHamiltonian)` that is really concise and doesn't print out all of the fcidump to make working on the REPL a bit nicer. Maybe printing the `starting_address` is sufficient?

It could also be nice if the type instance remembered which molecule this was for, maybe with a `specifier::String` field that could either hold the path to the FCIDUMP file or be a user-supplied name. Maybe in the future this could be uplifted from information in the `AtomsBase.jl` package.

joachimbrand reviewed (changes requested):
It is nice to see that you got iteration working without allocations. This is great progress!

I appreciate the comments and docstrings and improved readability of the iteration code. I think that the code can still be simplified, which can also make the code more readable and faster.

I'm not quite sure whether iteration works as it should, as repeatedly 'void' states are being returned (and I thought they shouldn't):
```julia-repl
julia> while result != nothing
       elem, state = result
       println(state)
       result = iterate(od, state)
       end
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (1, 0), (2, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (1, 0), (3, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (1, 0), (4, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (1, 0), (5, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (1, 0), (6, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (2, 0), (1, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (2, 0), (2, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (2, 0), (3, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (2, 0), (4, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (2, 0), (5, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (2, 0), (6, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 1), (0, 0), (0, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (1, 0), (2, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (1, 0), (3, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (1, 0), (4, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (1, 0), (5, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (1, 0), (6, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (2, 0), (1, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (2, 0), (2, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (2, 0), (3, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (2, 0), (4, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (2, 0), (5, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (2, 0), (6, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((1, 0), (0, 0), (0, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (1, 3))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (1, 4))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (1, 5))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (1, 6))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (2, 3))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (2, 4))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (2, 5))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (2, 6))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (3, 4))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (3, 5))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (3, 6))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (4, 5))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (4, 6))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (1, 2), (5, 6))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((0, 2), (0, 0), (0, 0))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((2, 0), (1, 2), (1, 3))
Rimu.Hamiltonians.MolecularHamiltonianOffDiagonalsIteratorState((2, 0), (1, 2), (1, 4))
...
```
This was for the `h2_dual` Hamiltonian and offdiagonals from the starting address.

lch committed:
Molecular Hamiltonian (#321)

1. Rename `ModeMap` to `FermiFS2CModes` and move to `BitStringAddresses` module.
2. Rename the fields of `MolecularHamiltonianOffDiagonalsIteratorState`.
3. Add docstring to several functions.

lch committed:
Molecular Hamiltonian (#321)

Enhances the documentation and readability of `MolecularHamiltonian`.

lch committed:
Molecular Hamiltonian (#321)

Improve the naming and document of `MolecularHamiltonian`.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -3,14 +3,20 @@
+        a^{\\dagger}_{j,\\sigma_{j}} a_{i,\\sigma_{i}} \\delta_{ij}, \\\\
+    \\hat{H}_2 & = \\sum_{\\sigma_k, \\sigma_l} \\sum_{\\sigma_i, \\sigma_j} \\sum_{kl,ij} 
+        V_{kl,ij} a^{\\dagger}_{k,\\sigma_k} a^\\dagger_{l,\\sigma_l} a_{j,\\sigma_j} a_{i,\\sigma_i} 
+        - V_{kl,ji} a^{\\dagger}_{k,\\sigma_k} a^\\dagger_{l,\\sigma_l} a_{i,\\sigma_i} a_{j,\\sigma_j} \\delta_{ij}.
```
Suggestion to simplify and make it more readable on the REPL by making use of unicode. The previous version was incorrect. Please double check that this is correct!
```suggestion
    Ĥ₁ & = \\sum_{i,j,σ} h_{ij}  a^†_{j,σ} a_{i,σ}, \\\\
    Ĥ₂ & = ½ \\sum_{i,j,σ,τ} V_{ijkl} a^†_{i,σ} a^†_{j,τ} a_{l,τ} a_{k,σ} .
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
````diff
@@ -3,14 +3,20 @@
 The expression of Hamiltonian under Bohn-Oppenheimer approximation is 
 ```math
 \\begin{aligned}
     \\hat{H}   & = \\hat{H}_0 + \\hat{H}_1 + \\hat{H}_2, \\\\
````
```suggestion

The expression for the Hamiltonian is 
```math
\\begin{aligned}
    Ĥ   & = h₀ + Ĥ₁ + Ĥ₂, \\\\
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
````diff
@@ -3,14 +3,20 @@
+        - V_{kl,ji} a^{\\dagger}_{k,\\sigma_k} a^\\dagger_{l,\\sigma_l} a_{i,\\sigma_i} a_{j,\\sigma_j} \\delta_{ij}.
 \\end{aligned}
 ```
 
````
```suggestion
where the constants ``h₀``, ``h_{ij}``, and ``V_{ijkl}`` are provided by the FCIDUMP `fd`. 
See the documentation of [ElemCo.jl](https://elem.co.il) for generating FCIDUMP information
with a Hartree-Fock calculation, or parsing them from a file.

```

joachimbrand reviewed (commented):
A few quick comments on the formula in the docstring. I see that some tests are failing so I leave that before I review the code again.

lch committed:
Molecular Hamiltonian (#321)

Add initial implement of iterator verion `random_offdiagonal` function.

lch committed:
Molecular Hamiltonian (#321)

Improve `MolecularHamiltonian` docstring

lch committed:
Molecular Hamiltonian (#321)

1. Fix broken tests
2. Add test for the functions supporting `random_offdiagonal`

lch committed:
Molecular Hamiltonian (#321)

1. Add example script describing the usage of `MolecularHamiltonian`.
2. Add missing `LOStructure()` for `MolecularHamiltonian`.
3. Add links to `MolecularHamiltonian` in the document of `Hamiltonian`
   module.

lch committed:
Molecular Hamiltonian (#321)

Optimize `unrank_combination` and add docstring for it.

lch committed:
Molecular Hamiltonian (#321)

Remove time consuming calculation in document.

lch committed:
Molecular Hamiltonian (#321)

Improve the document output.

joachimbrand commented on scripts/Molecule-example.jl:
```diff
@@ -0,0 +1,68 @@
+# # Example 6: Molecular Hamiltonian Caluculation
```
```suggestion
# # Example 6: Molecular Hamiltonian Calculation
```

joachimbrand commented on scripts/Molecule-example.jl:
```diff
@@ -0,0 +1,68 @@
+# Next we construct the Hamiltonian by the constructor. The Hartree-Fock ground state 
+# is generated automatically at the same time. 
+h = MolecularHamiltonian(fcidump)
+a = starting_address(h);
```
```suggestion
# Next we construct the Hamiltonian by the constructor [`MolecularHamiltonian`](@ref). 
# The Hartree-Fock ground state is generated automatically at the same time. 
h = MolecularHamiltonian(fcidump)
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,773 @@
+            starting_address::FermiFS2C,
+            specifier::String,
+        ]
+    )
```
```suggestion
    MolecularHamiltonian(fcidump_path::String[, starting_address::FermiFS2C]) <: AbstractHamiltonian
    MolecularHamiltonian(
        fd::ElemCo.FciDumps.FDump[, 
            starting_address::FermiFS2C,
            specifier::String,
        ]
    )
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,773 @@
+Implements an electronic ab-initio Hamiltonian based on electron overlap integrals in
+FCIDUMP format. It can be used to describe the electronic structure of a molecule
+with fixed nuclear positions (i.e. under the Born-Oppenheimer approximation).
+The parse of FCIDUMP file depends on [ElemCo.jl](https://elem.co.il). 
```
```suggestion
```
I suggest to delete this line as the information is redundant (provided later).

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,773 @@
+* `fd`: The FCIDUMP structure defined in [ElemCo.jl](https://elem.co.il).
+* `starting_address`: The starting address, defines number of particles and sites.
+* `specifier`: The string used to identify the Molecular Hamiltonian, it stores the path
+    to FCIDUMP file or a user specified name.
```
```suggestion
* `fcidump_path`: Path to the FCIDUMP file of overlap integrals. This can be generated with 
    standard quantum chemistry packages
* `fd`: Struct containing FCIDUMP information defined in [ElemCo.jl](https://elem.co.il).
* `starting_address`: The starting address (configuration) defines number of alpha and beta 
    electrons and orbitals.
* `specifier`: Arbitrary identifier. It may store the path to the FCIDUMP file or a user specified name.
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,773 @@
+    end
+
+    if num_modes(starting_address) != n_orb
+        throw(ArgumentError("starting_address provided must have same orbital numbers with FCIDUMP"))
```
```suggestion
        throw(ArgumentError("Starting_address must have the same number of orbital as the FCIDUMP."))
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,773 @@
+    address::A
+    op::H
+    modes::M
+    ind_cap::Tuple{Int,Int,Int,Int,Int}
```
Can we find a better name for this field that explains what is is about? Maybe `excitations_per_type`?

lch replied:
It is used to store prefix sums of number of excitation. I have also renamed it and added some explanation in docstring in the latest commit.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,773 @@
+function MolecularHamiltonianOffDiagonals(
+    c::MolecularHamiltonianOperatorColumn
+)
+    ind_cap_beta_1 = length(c.modes.occupied[2]) * length(c.modes.unoccupied[2])
+    ind_cap_alpha_1 = ind_cap_beta_1 + length(c.modes.occupied[1]) * length(c.modes.unoccupied[1])
+    ind_cap_beta_2 = ind_cap_alpha_1 + binomial(length(c.modes.occupied[2]), 2) *
+                                       binomial(length(c.modes.unoccupied[2]), 2)
+    ind_cap_alpha_2 = ind_cap_beta_2 + binomial(length(c.modes.occupied[1]), 2) *
+                                       binomial(length(c.modes.unoccupied[1]), 2)
+    ind_cap_alpha_1_beta_1 = ind_cap_alpha_2 + length(c.modes.occupied[1]) * length(c.modes.unoccupied[1]) *
+                                               length(c.modes.occupied[2]) * length(c.modes.unoccupied[2])
+    ind_cap = (ind_cap_beta_1, ind_cap_alpha_1, ind_cap_beta_2, ind_cap_alpha_2, ind_cap_alpha_1_beta_1)
```
The formatting of this code could be made a little prettier. Please keep to less than 92 characters per line! I think the internal variable names can be shortened, e.g. `beta_2` instead of `ind_cap_beta_2`. It is not obvious what `ind_cap` stands for, anyway.

lch replied:
I have formatted my code in the latest commit with [BlueStyle](https://github.com/JuliaDiff/BlueStyle) by [JuliaFormatter.jl](https://github.com/domluna/JuliaFormatter.jl).

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,773 @@
+    end
+end
+
+function Base.length(iter::MolecularHamiltonianOffDiagonals)
+    return length(iter.modes.occupied[1]) * length(iter.modes.unoccupied[1]) + # 1-alpha
+           length(iter.modes.occupied[2]) * length(iter.modes.unoccupied[2]) + # 1-beta
+           binomial(length(iter.modes.occupied[1]), 2) * 
+           binomial(length(iter.modes.unoccupied[1]), 2) + # 2-alpha
+           binomial(length(iter.modes.occupied[2]), 2) * 
+           binomial(length(iter.modes.unoccupied[2]), 2) + # 2-beta
+           (length(iter.modes.occupied[1]) * length(iter.modes.unoccupied[1]) *
+            length(iter.modes.occupied[2]) * length(iter.modes.unoccupied[2])) # 1-alpha, 1-beta
```
Isn't this information already contained in the `ind_cap` field and just needs to be added?

joachimbrand reviewed (commented):
Nice progress. I like the example script.
Please see a few comments below!

lch committed:
Molecular Hamiltonian (#321)

1. Improve the example documents.
2. Renaming fiels in `MolecularHamiltonianOffDiagonals` for clarity.

lch committed:
Molecular Hamiltonian (#321)

Format the code with BlueStyle (https://github.com/JuliaDiff/BlueStyle)

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,816 @@
+The `excitation_prefix_sums` field stores the prefix sums of excitation counts
+for each type, where each entry represents the cumulative number of excitations.
+Its length is fixed at 5. The ordering of excitation types must be consistent with
+the iteration order of `MolecularHamiltonianOffDiagonals`.
```
```suggestion
The iteration proceeds by going through the five possible excitation types:
1. single (particle - hole) excitations of the beta electrons
2. single (p-h) excitations of the alpha electrons
3. double (2p-2h) excitations of the beta electrons
4. double (2p-2h) excitations of the alpha electrons
5. mixed double excitations (ph of alpha and ph of beta electrons)

The field `excitation_prefix_sums` stores the sums of excitation counts for each 
of the five excitation types in order.

See also [`offdiagonals`](@ref).
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,816 @@
+"""
+This struct is used internally to represent the iterator for off-diagonal 
+terms generation, which is the returned value type of [`operator_column`](@ref) 
+function when applying on [`MolecularHamiltonian`](@ref). 
```
```suggestion
    MolecularHamiltonianOffDiagonals
This struct is used internally to represent the iterator for the generation
of off-diagonal terms from the [`operator_column`](@ref) of a
[`MolecularHamiltonian`](@ref). 
```

joachimbrand committed:
Make docstings of internal functions show up in Documenter docs

joachimbrand committed:
fix grammar

joachimbrand commented on scripts/Molecule-example.jl:
````diff
@@ -0,0 +1,68 @@
+p = ExactDiagonalizationProblem(h, algorithm=KrylovKitSolver(true))
+
+# Then it can be solved by executing
+# ```julia
+# s = solve(p)
+# ```
````
Maybe we should warn here that this may require larger computational resources.

I'm actually not sure what exactly happens and why this seems to stall. `solve(p; full_basis=true)` seems a good idea, as initialising the `LinearOperator` is done much faster (12s on my laptop), but the matrix-vector multiplications seem to stall. We can deal with this later, though, as I don't think it is an issue related to this PR.

joachimbrand commented on scripts/Molecule-example.jl:
```diff
@@ -0,0 +1,68 @@
+df = DataFrame(result);
+
+# To analyse the energy shift, we can use [`shift_estimator`](@ref).
+se = shift_estimator(df; skip=steps_equilibrate)
```
Add a bit of interpretation here, i.e. that the mean is an estimator for the energy. Also explain that and why you used the initiator and what this means for the energy estimate.

joachimbrand reviewed (commented):
A few more comments from me.

It would be good to have the unexported docstrings show up in Documenter as well.

Finally, please check the Codecoverage report, as parts of the Molecular.jl code are not tested currently. Please either add tests, or delete the code if it is obsolete!

lch committed:
Molecular Hamiltonian (#321)

1. Improve the document.
2. Remove obsolete code.

joachimbrand committed:
fix documentation

joachimbrand committed:
list unexported functions explicity in documentation

joachimbrand committed:
fix docs

joachimbrand committed:
include linked docs for FermiFS2CModes

lch committed:
fix docs

lch committed:
improve docs

lch committed:
fix tutorial output

lch committed:
Molecular Hamiltonian (#321)

Make ElemCo an optional dependency through a package extension.

lch committed:
fix docs

lch committed:
add mising dependency

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+See also [`FermiFS2C`](@ref)
+
+!!! warning
+    The FCIDUMP file is assumed to be a Restricted Hartree-Fock(RHF) Hamiltonian.
```
```suggestion
!!! note
    The FCIDUMP file is assumed to be a Restricted Hartree-Fock(RHF) Hamiltonian.
!!! note
    Requires the [ElemCo.jl](https://elem.co.il) package to be loaded with `using ElemCo`.
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+    fcidump_path::String,
+    starting_address::Union{Nothing,FermiFS2C}=nothing,
+    specifier::String="",
+)
```
The string that is printed at the REPL currently does not parse as a valid constructor:
```julia-repl
julia> h = MolecularHamiltonian(fcidump)
MolecularHamiltonian("/Users/brand/git/code/Rimu/test/examples/h2o.FCIDUMP", starting_addresss=fs"|⇅⇅⇅⇅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⟩")

julia> h == MolecularHamiltonian("/Users/brand/git/code/Rimu/test/examples/h2o.FCIDUMP", starting_addresss=fs"|⇅⇅⇅⇅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⋅⟩")
ERROR: MethodError: no method matching MolecularHamiltonian(::String; starting_addresss::FermiFS2C{4, 4, 23, 8, FermiFS{…}, FermiFS{…}})
This method may not support any kwargs.

Closest candidates are:
  MolecularHamiltonian(::String) got unsupported keyword argument "starting_addresss"
   @ Rimu ~/git/code/Rimu/src/Hamiltonians/Molecular.jl:46
  MolecularHamiltonian(::String, ::Union{Nothing, FermiFS2C{N1, N2, M} where {N1, N2, M}}) got unsupported keyword argument "starting_addresss"
   @ Rimu ~/git/code/Rimu/src/Hamiltonians/Molecular.jl:46
  MolecularHamiltonian(::String, ::Union{Nothing, FermiFS2C{N1, N2, M} where {N1, N2, M}}, ::String) got unsupported keyword argument "starting_addresss"
   @ Rimu ~/git/code/Rimu/src/Hamiltonians/Molecular.jl:46
  ...
```
This can be fixed by making the optional arguments keyword arguments:
```suggestion
function MolecularHamiltonian(
    fcidump_path::String;
    starting_address::Union{Nothing,FermiFS2C}=nothing,
    specifier::String="",
)
```
If you make this change, please update the docstring such that it is accurate!
I'd also like to see this tested, e.g. by adding something like this to the tests:
```julia
            @test eval(Meta.parse(repr(H))) == H
```
Obviously, this only works when constructing from a path, and when the path is still available and accurate. 

Update: I notice that for this to work you may need to define a method for `isequal` for the `TFDump` struct. This can be done in the package extension (or ideally as a PR to ElemCo.jl, as otherwise it would be type piracy 😄) .

lch replied:
Maybe we could mark the test to be skipped[^skip].
```julia
he = eval(Meta.parse(repr(h)))
@test he == h skip=true
```
Then we wait for the changes to be deployed by ElemCo.jl.

[^skip]: As document of [`@test`](https://docs.julialang.org/en/v1/stdlib/Test/#Test.@test) suggests, test with `skip=true` could be used to mark "tests of not-yet-implemented functionality".

joachimbrand replied:
Yes, good idea!

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+        occ_modes::Tuple{AbstractVector{FermiFSIndex},AbstractVector{FermiFSIndex}}
+    )::T where {T<:Number}
+
+Calculate the one body operator diagonal term ``\\langle U | \\hat{H}_1 | U \\rangle``.
```
```suggestion
    one_electron_diagonal(
        int1::Matrix{<:Number},
        occ_modes::NTuple{2,<:AbstractVector{FermiFSIndex}}
    )

Calculate the one body operator diagonal term ``⟨U|Ĥ₁|U⟩`` for a two-component Fermi Fock address.
```
I suggest to rename the function to make it clearer what the function does.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+        \\langle \\varphi_i | a^{\\dagger}_{i} a_{i} | \\varphi_i \\rangle
+      + \\sum_{i=1}^{n_{\\beta}} h_{ii}
+        \\langle \\varphi_i | a^{\\dagger}_{i} a_{i} | \\varphi_i \\rangle.
+\\end{aligned}
```
```suggestion
⟨U|Ĥ₁|U⟩ = ∑_{i,σ}^N ⟨U| h_{ii} a^†_{iσ} a_{iσ} |U⟩ ,
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
````diff
@@ -0,0 +1,780 @@
+        \\langle \\varphi_i | a^{\\dagger}_{i} a_{i} | \\varphi_i \\rangle.
+\\end{aligned}
+```
+"""
````
```suggestion
where the mode map of the two-component Fock address ``|U⟩`` is passed as `occ_modes`.
"""
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+end
+
+"""
+    two_electron_integral(
```
```suggestion
    two_electron_diagonal(
```
I suggest to rename this function as well. Please rewrite the rest of the docstring in line with the changes suggested for `one_electron_diagonal`!

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+* `from_occupieds`: Records the indices of `FermiFS2CModes` arrays which modes electrons
+being excited from.
+* `to_unoccupieds`: Records the indices of `FermiFS2CModes` arrays which modes electrons
+being excited to.
```
```suggestion
* `excitations_per_channel`: Tuple contains 2 values represents the how many electrons
    are excited in alpha(1) and beta(2) channel.
* `from_occupieds`: Records the indices of `FermiFS2CModes` arrays which modes electrons
    being excited from.
* `to_unoccupieds`: Records the indices of `FermiFS2CModes` arrays which modes electrons
    being excited to.
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+
+Both `from_occupieds == (0,0)` and `to_unoccupieds == (0,0)` represent a special "void"
+state when there is no more state in current `n_excited` situation, this is used to
+notify upper caller that it should move to next valiad `excitations_per_channel`.
```
```suggestion
notify upper caller that it should move to next valid `excitations_per_channel`.
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+When used to represents two-electrons-excitation cases, `from_occupieds[1]`,
+`to_unoccupieds[1]`, `from_occupieds[2]` and `to_unoccupieds[2]` are all used.
+When used to represents one-one-electrons-excitation case, index `1` represention
+alpha spin channel and index `2` represention beta spin channel repectively.
```
```suggestion
When used to represent the one-one-electron-excitation case, index `1` represents
alpha electrons and index `2` represents beta electrons, respectively.
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+    il::Int,
+)
+    return MolecularHamiltonianOffDiagonalsIteratorState(n_excited, (ii, ij), (ik, il))
+end
```
Please move the constructors right after the type definition! This makes it easier to read the code.

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+- Since the number of states for each excitation type is known in advance,
+  we can determine which excitation type corresponds to the given `iter_index`
+  by checking the index range it falls into.
+
```
```suggestion
```
I think that you may need to delete this empty line for the markdown counter to continue. Please check what the web documentation looks like!

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+represent modes the electrons should go to as `to_unoccupieds` tuple in
+[`MolecularHamiltonianOffDiagonalsIteratorState`](@ref).
+If there are 4 unoccupied modes, all the possible combination of chosen in
+lexicographic order should be:
```
```suggestion
Return the 2-element combination (in lexicographical order) of the set ``{1,...,n}`` with index ``i``.
When generating the excitation state with 2 electrons excited in a single spin
channel, we use the indices of the array storing the (un)occupied mode maps. For example,
we select 2 non-repeating indices of array storing unoccupied mode maps to
represent the modes the electrons should go to as  the`to_unoccupieds` tuple in
[`MolecularHamiltonianOffDiagonalsIteratorState`](@ref).
If there are 4 unoccupied modes, all the possible combination of chosen tuples in
lexicographic order should be:
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,780 @@
+The length of all the combinations is ``4C2 = 6``.
+Given the index(`i`), to determin the value of 1st position(`x`) in the
+combination, we can first determine how many combinations when `x` is
+from 1 to `n`, the number is calculated by ``(n-x)C1 = (n-x)``.
```
```suggestion
The length of all the combinations is ``\\binom{4}{2} = 6``.
Given the index(`i`), to determine the value of the first position(`x`) in the
combination, we can first determine the number of combinations when `x` goes
from 1 to `n`. The number is calculated by ``\\binom{n-x}{1} = n-x``.
```
You'll need to check whether the `amsmath` package of latex is supported. Otherwise you can leave it as you had it.

joachimbrand commented on test/molecular_hamiltonian_tests.jl:
```diff
@@ -0,0 +1,318 @@
+using Rimu
+using Test
+using Rimu.InterfaceTests: test_observable_interface, test_operator_interface,
+    test_hamiltonian_interface, test_hamiltonian_structure
```
```suggestion
using Rimu.InterfaceTests: test_hamiltonian_interface, test_hamiltonian_structure
```

joachimbrand reviewed (changes requested):
Nice progress 👍 

The package extension mechanism seems to work well.
See comments below, mostly for readability and documentation.
This should be ready to be merged any time.
Don't forget to change the status from Draft to Ready!

lch committed:
Molecular Hamiltonian (#321)

1. Improve `MolecularHamiltonian` constructor.
2. Rename functions calculating diagonal term.
3. Fix document display error.
4. Improve imath equation for clarity under REPL.
5. Improve method `unrank_combination` document.

lch committed:
correct function name in docs

lch committed:
Molecular Hamiltonian (#321)

1. Fix the test of `MolecularHamiltonian`.
2. Fix the typo in documents.

lch committed:
fix docs

joachimbrand commented on src/Hamiltonians/Molecular.jl:
````diff
@@ -0,0 +1,790 @@
+
+Calculate the two body operator diagonal term ``⟨U|Ĥ₂|U⟩`` for a two-component Fermi Fock address.
+```math
+⟨U|Ĥ₂|U⟩ 
+    = ∑_{i < j,σᵢ,σⱼ}^{N} (V_{ij,ij} a^†_{i,σᵢ} a^†_{j,σⱼ} a_{j,σⱼ} a_{i,σᵢ} 
+    - V_{ij,ji} a^†_{i,σᵢ} a^†_{j,σⱼ} a_{i,σᵢ} a_{j,σⱼ} δ_{σᵢ,σⱼ})
````
This formula is not correct as the LHS is a number and the RHS an operator. Also, I don't understand why the spins should be restricted to be the same in the second term. Should it not be the following?
```suggestion
⟨U|Ĥ₂|U⟩ 
    = ∑_{i < j,σ,τ}^{M} (V_{ij,ij} ⟨U| a^†_{i,σ} a^†_{j,τ} a_{j,τ} a_{i,σ} |U⟩ 
    - V_{ij,ji} ⟨U| a^†_{i,σ} a^†_{j,τ} a_{i,τ} a_{j,σ}|U⟩ )
```
NB: The sum go over `M` orbitals.

lch replied:
The two-body operator term for occupied orbitals `i` and `j` is given by `J_ij - K_ij`, where `J` is the Coulomb integral and `K` is the exchange integral. The factor `δ_{σ,τ}` is required because when `i` and `j` belong to different spin channels, the exchange integral vanishes, which is reflected in the code.
```julia
sum_alpha_beta = zero(T)
for i in occ_modes[1]
    for j in occ_modes[2]
        sum_alpha_beta += int2[i.mode, j.mode, i.mode, j.mode]
    end
end
two_elec_int += sum_alpha_beta
```

joachimbrand replied:
OK, thanks!

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,790 @@
+\\begin{aligned}
+    Ĥ  & = h₀ + Ĥ₁ + Ĥ₂, \\\\
+    Ĥ₁ & = \\sum_{i,j,σ} h_{ij}  a^†_{j,σ} a_{i,σ}, \\\\
+    Ĥ₂ & = ½ \\sum_{i,j,σ,τ} V_{ijkl} a^†_{i,σ} a^†_{j,τ} a_{l,τ} a_{k,σ}.
```
I still found mistakes in the formula.
```suggestion
    Ĥ₁ & = \\sum_{i,j,σ} h_{ij}  a^†_{i,σ} a_{j,σ}, \\\\
    Ĥ₂ & = ½ \\sum_{i,j,k,l,σ,τ} V_{ijkl} a^†_{i,σ} a^†_{j,τ} a_{l,τ} a_{k,σ}.
```
Please double check that they are correct. It is very confusing to have incorrect formulas in the documentation.

lch replied:
Maybe we should align the notation of subscript with indices of `fcidump.int2` and mode indices of `excitation` used in the code. I have updated it in the latest commit for a test.

joachimbrand replied:
OK.

joachimbrand commented on scripts/Molecule-example.jl:
```diff
@@ -0,0 +1,72 @@
+# [here](https://github.com/RimuQMC/Rimu.jl/blob/develop/scripts/Molecule-example.jl).
+# Run it with `julia Molecule-example.jl`.
+
+# First, we load the reqired packages. `Rimu` for the calculation and `ElemCo` for loading the FCIDUMP file.
```
```suggestion
# First, we load the required packages. `Rimu` for the calculation and `ElemCo` for loading the FCIDUMP file.
```

joachimbrand commented on src/Hamiltonians/Molecular.jl:
```diff
@@ -0,0 +1,790 @@
+    if ispath(h.specifier)
+        print(io, "\"", h.specifier, "\"")
+    else
+        print(io, typeof(h.fcidump))
```
```suggestion
        print(io, "::", typeof(h.fcidump))
```
If you try to type it into the repl, you'll get an error message either way. But indicating that this is a typename may clarify why that is.

joachimbrand reviewed (changes requested):
We need to get the formulas right in the documentation. Otherwise this is looking good.

lch committed:
Molecular Hamiltonian (#321)

Improve the docstring of `MolecularHamiltonian`

joachimbrand reviewed (approved):
Thanks for the latest updates and for the big effort that went into this project! The PR looks good now and is ready be merged 🚀
